### PR TITLE
perlxstut add links and update examples

### PR DESCRIPTION
**Pull request by ghost**
For https://github.com/Perl/perl5/issues/18163.

### CODE REVIEW

**Comment by apparluk** [comment]
@xsawyerx LGTM, if you want to merge.

**Review by atoomic — APPROVED** [review]
LGTM

**Comment by apparluk** [comment]
proposed comments not added.owing to line length

>
>```
>#define PERLIO_NOT_STDIO 0      /* For co-existence with stdio only */
>#define PERL_NO_GET_CONTEXT     /* we want efficiency */
>```

**Comment by karenetheridge** [comment]
Why is perl 5.22.2 needed?  Could the commit message be a little more self-explanatory?

**Comment by apparluk** [comment]
@karenetheridge factually 5.22.2 was mentioned at the top of the ticket.

**Comment by apparluk** [comment]
> Why is perl 5.22.2 needed?

It's just what I was running when I tested the old examples.

The change to Mytest.pm can be reverted or substituted for another Perl version according to developer preferences.


```diff
-    use 5.008008;
+    use 5.022002;
```

I thought the `h2xs` output files' renderings in the tutorial should reflect currently working examples.

**Comment by apparluk** [comment]
> Why is perl 5.22.2 needed?

Factually, I've reverted that now in the .pod file.

**Comment by apparluk** [comment]
@toddr do you want to go ahead and merge?

**Comment by apparluk** [comment]
@karenetheridge can you merge?

**Comment by apparluk** [comment]
@khwilliamson can you merge?

The modified examples all work as expected.

**Review by xsawyerx — APPROVED** [review]

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (Grinnz) [inline dist/ExtUtils-ParseXS/lib/perlxstut.pod:125]: I still do not see why this is being modified.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (Grinnz) [inline dist/ExtUtils-ParseXS/lib/perlxstut.pod:610]: Nor this.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (Grinnz) [inline dist/ExtUtils-ParseXS/lib/perlxstut.pod:625]: Why make this less explanatory? This is referring to the initial layout created by h2xs.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (Grinnz) [inline dist/ExtUtils-ParseXS/lib/perlxstut.pod:1408]: Why remove this email?

**Comment by Grinnz** [comment]
In general it is counter-productive to include unrelated unexplained changes.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (ghost) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:1408]: The author's email address was outdated.

I notified the individual of its removal and was returned a thank you.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (ghost) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:125]: EXAMPLE 1

I take the presumption here that to not include `-I.` would constitute an editing of the `h2xs -A -n Mytest` generated Makefile.PL; which would then perhaps oblige a justification presented to the reader as to the purpose for that editing.

[What has gone on?](https://perldoc.perl.org/perlxstut#What-has-gone-on?)

...In later examples we'll see how we can use h2xs to read header files and generate templates to connect to C routines.

<hr>

[re::engine::Plan9](https://metacpan.org/release/re-engine-Plan9) (Mar 31, 2010) for example includes an .h file in the top level extracted download

Comment (Grinnz) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:125]: I see. The explanation for this since it has not been mentioned is that h2xs now generates a dist with ppport.h causing it to include `-I.` in the generated Makefile.PL.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (Grinnz) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:625]: The `mylib` is no longer detected by the initial `make manifest`. I think the first sentence should be kept, and the second sentence can be replaced by "by appending the following three lines:"

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (Grinnz) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:610]: Resolved as the above comment.

Comment (ghost) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:610]: EXAMPLE 4

The tutorial instructs to add in Makefile.PL an argument to the WriteMakefile call and to add a subroutine.

I take the presumption here that further editing of the WriteMakefile call would perhaps oblige justification.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (ghost) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:625]: EXAMPLE 4

`% h2xs -O -n Mytest2 Mytest2/mylib/mylib.h`

The above command generates a MANIFEST file which does not include a "mylib" line.

**Comment by apparluk** [comment]
@xsawyerx and others my apologies if I appeared impatient to push forward a merge.

GitHub updated and no longer supports the browser I typically use on an old Netbook.

> In general it is counter-productive to include unrelated unexplained changes.

@Grinnz I certainly concur on that point.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (ghost) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:625]:
```diff
  Let's also fix the MANIFEST file so that it accurately reflects the contents
- of our extension.  The single line that says "mylib" should be replaced by
+ of our extension. by appending the following three lines:
- the following three lines:
```

That looks awkward to me.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (karenetheridge) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:610]: *all* changes oblige justification, and should be grouped into separate pull requests when unrelated.

**Comment by apparluk** [comment]
Was `#define PERL_NO_GET_CONTEXT` added to `h2xs.PL` later than 5.8.x?

**Comment by ghost** [comment]
The define `PERL_NO_GET_CONTEXT` was added to `h2xs.PL`on Dec 6, 2012. 

The define `PERL_NO_GET_CONTEXT` was added to ParseXS's pods on Feb 16, 2014.

Perl `5.008008`  could have been updated to `5.017007` concomitant with either of those commits.

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (ghost) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:625]: Would this be more to your sympathies?

Let's also fix the MANIFEST file by running the following command in the top level directory.

```
% perl -MExtUtils::Manifest=mkmanifest -e 'mkmanifest()'
```

Comment (Grinnz) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:625]: Sorry I did not mean literally replace the second sentence. "Let's also fix the MANIFEST file so that it accurately reflects the contents of our extension by ..."

Running mkmanifest is a better suggestion than manually adding the lines, this could also be done at the end after a Makefile is generated with `make manifest`.

**Comment by apparluk** [comment]
EXAMPLE 1

The `h2xs -A -n Mytest` generated Mytest.xs content as presented was not possible prior to 5.17.7.

Could an organization Member present an opinion on https://github.com/Perl/perl5/pull/18240#issuecomment-719445846?

File: dist/ExtUtils-ParseXS/lib/perlxstut.pod
Comment (apparluk) [reply dist/ExtUtils-ParseXS/lib/perlxstut.pod:625]: No apprehension about introducing a one-liner in a .pod file otherwise free of them?

> Why make this less explanatory?

The command output would be self-explanatory. 

So you want the below change?

Let's also fix the MANIFEST file by running the following command in the top-level directory.

```
% perl -MExtUtils::Manifest=mkmanifest -e 'mkmanifest()'
```

If you want the change I'll await a response on https://github.com/Perl/perl5/pull/18240#issuecomment-719833092 before adding this change.

**Comment by ghost** [comment]
3 failing checks after https://github.com/Perl/perl5/pull/18240/commits/bb321ad87ad579a8bd2037cdf9b813019a1e25a9.

testsuite / Windows msvc142
testsuite / Windows msvc100
testsuite / Windows mingw64

**Comment by xsawyerx** [comment]
@apparluk May I ask why you closed this PR and deleted the branch?

**Comment by ghost** [comment]
@xsawyerx 

[doc] perlxs rpc (also Closed)

Here are the two files if someone else wants to take up the work.

Gist <s>perlxstut.pod</s> (<s>raw</s>) [deleted Nov 5, 2020]

Contains your code changes (the last all tests passing) and `use 5.017007;` but not `mkmanifest`.

Gist <s>perlxs.pod</s> (<s>raw</s>) [deleted Nov 5, 2020]

Run rpctest.pl and verify the correctness of the output before merging.

<hr>

I have mostly moved on from this document. Please feel free to make whatever changes you feel are necessary to keep it up to date, or expand it as much as you like.

**Comment by khwilliamson** [comment]
#18278 has been opened to continue work on xsxtut.pod
#18279 has been opened to continue work on perlxs

**Comment by ghost** [comment]
@khwilliamson 

cf: https://github.com/Perl/perl5/pull/18240#issuecomment-719445846

I used 5.017007 in the .pod believing PERL_NO_GET_CONTEXT first appeared in that release.

```diff
-   use 5.008008;
+   use 5.017007;
```

However it would be unusual to expose the reader to a development release of perl in the .pod, and maybe using instead the >5.017007 next stable perl version would be more sensible.

<hr>

_cf_: https://github.com/Perl/perl5/pull/18240#issuecomment-720021784

The first failed tests on this PR occurred after https://github.com/Perl/perl5/pull/18240/commits/bb321ad87ad579a8bd2037cdf9b813019a1e25a9 (Nov 1, 2020, 2:31 AM GMT+1).

**Comment by ddobbelaere** [comment]
The failing tests are unrelated to the commits in this PR, see #18280

**Comment by ghost** [comment]
This PR's msvc fails had the error message citing an unresolved external symbol.

**Comment by xsawyerx** [comment]
> @xsawyerx
>
> [...] 
> 
> Many contributors/members are from the USA.
> 
> In the 1970s there was a deemphasis on the study of humanities in universities in the USA.
> 
> I find it difficult to collaborate with people who I view as close-minded.

I'm sorry this was your experience while trying to contribute to Perl. I understand the review was rigorous and that was probably mentally and emotionally exhausting.

With Karl moving your work to a new PR, I see us merging your improvements to the tutorial. I hope that in the future we could do better when reviewing by having more empathy for a contributor and making it possible to merge changes without jumping through endless hoops. It isn't usually the case, and I hope it won't be the case for you in the future.

One practical suggestion I have is to suggest breaking things up to multiple smaller PRs that allow simpler review and merge process, thus allowing you to build on top of approved changes each time.

**Comment by ghost** [comment]
Keep the append instruction and leave `mkmanifest` for another PR.

SPECIAL NOTES

Possible expansion could add a subsection on `make manifest` with mention of `./Build manifest`.

**Comment by apparluk** [comment]
@xsawyerx

Perl 5.16.0 (released 2012-05-20)

[Anatomy of a typemap](https://perldoc.perl.org/5.20.0/perlxstypemap#Anatomy-of-a-typemap)

"typemaps can also be embedded directly into XS code"

The ext subdirectory typemap files do not seem to be frequently modified.

A quick assessment could be made on embedding the typemaps in *.xs files.

If it looks favourable to pursue embedding then a new issue could be opened.  

<hr>

perl5/ext/Amiga-ARexx/
amigaos4: add Amiga::ARexx and Amiga::Exec 5 years ago

perl5/ext/Amiga-Exec/
amigaos4: add Amiga::ARexx and Amiga::Exec 5 years ago

perl5/ext/B/
Prepare B typemap for upcoming pad name changes 6 years ago

perl5/ext/GDBM_File/
Eliminate the fifth argument from GDBM_File::TIEHASH. 8 years ago

perl5/ext/NDBM_File/
Eliminate the fifth argument from GDBM_File::TIEHASH. 8 years ago

perl5/ext/ODBM_File/
ODBM: fix a compiler warning. 6 years ago

perl5/ext/POSIX/
POSIX: Check fds against negatives. 5 years ago

perl5/ext/SDBM_File/
Eliminate the fifth argument from GDBM_File::TIEHASH. 8 years ago

perl5/ext/XS-APItest/
APItest: Add U8* typemap, and use it 3 years ago

**Comment by apparluk** [comment]
@xsawyerx 

I find both https://github.com/Perl/perl5/issues/18163#issuecomment-703325290 and https://github.com/Perl/perl5/pull/18240#issuecomment-720988032 inappropriate.

Respectfully I'd request that you delete them and not post a response to this current comment..

**Comment by karenetheridge** [comment]
@apparluk please see https://perldoc.perl.org/perlpolicy#STANDARDS-OF-CONDUCT for our community standards. Feel free to contact me off-github to discuss this (I'm also on irc.perl.org as "ether").

**Pull request closed without merging**
